src/core/file.cc

```diff
@@ -381,6 +381,35 @@ std::optional<directory_entry_type> dirent_type(const linux_dirent64& de) {
     return type;
 }
 
+#ifdef SEASTAR_COROUTINES_ENABLED
+static coroutine::experimental::generator<directory_entry, dir_entry_buffer> make_list_directory_generator(coroutine::experimental::buffer_size_t buffer_size, thread_pool& pool, int fd) {
+    temporary_buffer<char> buf(8192);
+
+    while (true) {
+        auto size = co_await engine().readdir(fd, buf.get_write(), buf.size());
+        if (size == 0) {
+            co_return;
+        }
+
+        for (const char* b = buf.get(); b < buf.get() + size; ) {
+            const auto de = reinterpret_cast<const linux_dirent64*>(b);
+            b += de->d_reclen;
+            sstring name(de->d_name);
+            if (name == "." || name == "..") {
+                continue;
+            }
+            std::optional<directory_entry_type> type = dirent_type(*de);
+            directory_entry ret(std::move(name), type);
+            co_yield ret;
+        }
+    }
+}
+
+coroutine::experimental::generator<directory_entry, dir_entry_buffer> posix_file_impl::experimental_list_directory() {
+    return make_list_directory_generator(coroutine::experimental::buffer_size_t{128 / sizeof(directory_entry)}, *engine()._thread_pool, _fd);
```

Comment:
> sizeof(seastar::directory_entry) is 24. i am not able to find any clues on where 128 comes from.

The buffer-size argument here is the number of entries, not the number of bytes, isn't it?

Reply:
yeah, number of elements, the former.

Reply:
probably i failed to explain my question. let me rephrase: where does the 128 in "128 / sizeof(directory_entry)" come from?

Reply:
From ... dunno ... pretty much nowhere, just a buffer that's larger than a single entry, but still small enough. And the power-of-two of course

Reply:
BTW, name on directory entry is sstring, not online buffer, so it doesn't contribute to generator buffer capacity

Reply:
> From ... dunno ... pretty much nowhere, just a buffer that's larger than a single entry, but still small enough. And the power-of-two of course

I think @tchaikov's point is that it's a magic number, not that it's a wrong number.
Whatever is the reason for choosing 128, it should be explained with a comment it the code.

Reply:
Both, I also suggest another magic number in the first comment in this thread.

Reply:
> BTW, name on directory entry is sstring, not online buffer, so it doesn't contribute to generator buffer capacity

Hmm, the name string is returned by the provided buffer. I double checked by reading the Linux source.

Reply:
The generator buffer is used to keep seastar dirents in, not linux dirents

Reply:
> Whatever is the reason for choosing 128, it should be explained with a comment in the code.
> ...
> Both, I also suggest another magic number in the first comment in this thread.

Magic of 128 is the same as magic of 8192 to me :) From the "no magic" perspective it better be a generator with std::optional<> buffer type feeding one dirent at a time. No magic here, but not nice either

Reply:
I think that the source of confusion is that there are _two_ buffers.

One is used to get linux dirents from the kernel. This buffer is 8k in size and it's just the same as in legacy list_directory() method. This buffers fits in dirents _and_ their names. It's "that large" to make less (pretty expensive) system calls.

The other one is the generator buffer used by co_yield code to accumulate several yielded elements before switching between coroutines. It's 128 bytes in size, not 8k, because it's used to keep _only_ 24-bytes seastar dirents, since their names are sstrings and are allocated _elsewhere_. This buffer doesn't need to be 8k as it doesn't lead to more system calls, only coroutines switches.

Reply:
i see. sorry for the noise. and thank you for the detailed explanations.